watch-parties/src/main/java/com/google/sps/servlets/ChatListerServlet.java

```diff
@@ -45,37 +44,37 @@
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
     // Bring comments back from datastore
-    Query query = new Query("Chat").addSort("timestamp", SortDirection.DESCENDING); 
+    Query query = new Query("Chat").addSort("timestamp", SortDirection.ASCENDING); 
```

Comment:
I think I may have missed the point of the stack you were using previously. You need the 50 most recent chat messages (which you were fetching by sorting by timestamp, descending order), but you need to _serve_ them in ascending timestamp order. The stack took care of that, because stacks are LIFO (last in, first out).

What we really want is the 50 most recent chat entries. This is easiest to get if we sort in descending timestamp order.
```java
Query query = new Query("Chat").addSort("timestamp", SortDirection.DESCENDING);
```
Next, we can limit the number of results that we get back from Datastore (since we only want to serve 50, we just ask for 50).
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/datastore/FetchOptions.Builder.html#withLimit-int-
```java
PreparedQuery results = datastore.prepare(query);
List<Entity> entities = results.asList(FetchOptions.Builder.withLimit(50));
```
I also suggest getting the results as a `List` in this case (instead of an `Iterable`) because we want to iterate through it backwards (which is must easier to do with a `List`).
Iterating through the list backwards lets us serve entries that were sorted reverse chronologically (descending timestamp) in forward chronological order.
```java
for (int i = entities.size() - 1; i >= 0; i--) {
  Entity entity = entities.get(i);
  //
}
```
This lets us get the 50 most recent chat entries, served in chronological order.

Reply:
I suggest this only because the number of chat messages that could exist in a Room is unbounded, and we don't want to have to iterate through all of them if we can help it. In practice, what you have here would work just fine for a long time. The changes I'm suggesting are just some future-proofing (and good engineering).

Reply:
Uuuhhhh I mean "what Chy said" 😅
(I didn't see his comments when I started writing this)

Reply:
+1 to this. Let us know if you want talk through this over GVC.